packages/slate-dom/src/utils/dom.ts

```diff
@@ -355,3 +355,79 @@ export const isAfter = (node: DOMNode, otherNode: DOMNode): boolean =>
     node.compareDocumentPosition(otherNode) &
       DOMNode.DOCUMENT_POSITION_FOLLOWING
   )
+
+/**
+ * Shadow DOM-aware version of Element.closest()
+ * Traverses up the DOM tree, crossing shadow DOM boundaries
+ */
+export const closestShadowAware = (
+  element: DOMElement | null | undefined,
+  selector: string
+): DOMElement | null => {
+  if (!element) {
+    return null
+  }
+
+  let current: DOMElement | null = element
+
+  while (current) {
+    // Check if current element matches the selector
+    if (current.matches && current.matches(selector)) {
+      return current
+    }
+
+    // Try to go to the parent element
+    if (current.parentElement) {
+      current = current.parentElement
+    } else if ((current.parentNode as any)?.host) {
```

Comment:
Do these two any types need to be any, or can we make the type more specific?

Reply:
Sorry, leftover. Added proper types in [c7b8139](https://github.com/ianstormtaylor/slate/pull/5963/commits/c7b8139762a5ac45b91900dac0de6b788905cb41)